### PR TITLE
Issue-82: Allow configuring default scopes per client

### PR DESCRIPTION
**Pull request by vroldanbet**
- all POJOs are now using Lombok

### CODE REVIEW

**Comment by hjacobs** [comment]
@vroldanbet I rather have two different commits: one for Lombok and the other for the actual change (just FYI now).

Please also fix the decreasing code coverage.

**Comment by vroldanbet** [comment]
@hjacobs, that will be the case, two separated commits. I tried to minimize the decrease of code coverage by keeping generated code to minimal (most objects are now simple immutable + builder pattern). However, I fail to understand what is not being tested anymore according to CodeCov report.

**Comment by danieldelhoyo** [comment]
+

File: src/main/java/org/zalando/planb/provider/ScopeService.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    private static final String SPACE = " ";
+
+    @Autowired
+    private ScopeProperties scopeProperties;
+
+    public Set<String> getDefaultScopesForClient(final ClientRealm clientRealm, final String clientId) {
+        return clientRealm
+                .get(clientId)
+                .map(ClientData::getDefaultScopes)
+                .filter(t -> !t.isEmpty())
```
Comment (danieldelhoyo) [inline src/main/java/org/zalando/planb/provider/ScopeService.java]: another option here would be to create a function for negating a predicate and use it, sth like:
`    private static <T> Predicate<T> not(Predicate<T> t) {
        return t.negate();
    }`
and then the filter would be
`not(Set::isEmpty)`
but maybe like it is right now it's just easier to read, not sure tbh :smile:

**Comment by danieldelhoyo** [comment]
+1

File: src/main/java/org/zalando/planb/provider/ScopeService.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+public class ScopeService {
+
+    private static final String SPACE = " ";
+
+    @Autowired
+    private ScopeProperties scopeProperties;
+
+    public Set<String> getDefaultScopesForClient(final ClientRealm clientRealm, final String clientId) {
+        return clientRealm
+                .get(clientId)
```
Comment (hjacobs) [inline src/main/java/org/zalando/planb/provider/ScopeService.java]: This triggers an extra Cassandra call, I think we should only do one call to load client data..

File: src/main/java/org/zalando/planb/provider/ScopeService.java (hunk 1 above)
Comment (vroldanbet) [inline src/main/java/org/zalando/planb/provider/ScopeService.java]: That would be nice to improve readability @danieldelhoyo, but I rather not create another method call just to negate :dancer:

File: src/main/java/org/zalando/planb/provider/ScopeService.java (hunk 2 above)
Comment (vroldanbet) [inline src/main/java/org/zalando/planb/provider/ScopeService.java]: Good point @hjacobs, will look into it.

Comment (vroldanbet) [inline src/main/java/org/zalando/planb/provider/ScopeService.java]: Got rid of one unnecessary Cassandra call. The other one in org.zalando.planb.provider.OIDCController#createToken is unavoidable at this point.

File: src/main/java/org/zalando/planb/provider/ScopeService.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Autowired
+    private ScopeProperties scopeProperties;
+
+    public Set<String> getDefaultScopesForClient(final ClientRealm clientRealm, final ClientData clientData) {
+        return clientData.getDefaultScopes().isEmpty() ?
+                getDefaultScopesByRealm(clientRealm.getName()) : clientData.getDefaultScopes();
+    }
+
+    public Set<String> getDefaultScopesForClient(final ClientRealm clientRealm, final String clientId) {
```
Comment (hjacobs) [inline src/main/java/org/zalando/planb/provider/ScopeService.java]: now you can call the upper method here (to be more DRY)

Comment (vroldanbet) [inline src/main/java/org/zalando/planb/provider/ScopeService.java]: done :)

**Comment by danieldelhoyo** [comment]
+1

**Pull request merged**